### PR TITLE
StaticTokenCredential

### PR DESCRIPTION
**Pull request by christothes**
#19404
#21940
#19443

### CODE REVIEW

**Comment by pakrym** [comment]
What about a static method? `TokenCredential.Create(string token)`?

**Comment by christothes** [comment]
> What about a static method? `TokenCredential.Create(string token)`?
> 
> We did a similar thing with `Page` and `AsyncPageable`.

Instead of a ctor?

**Comment by pakrym** [comment]
>Instead of a ctor?

Instead of an additional class. TokenCredential.Create(string token) would return a `TokenCredential`

**Comment by christothes** [comment]
> > Instead of a ctor?
> 
> Instead of an additional class. TokenCredential.Create(string token) would return a `TokenCredential`

Interesting idea - but I'm not sure that I like having a generic TokenCredential that can't be identified specifically as a static token credential.

**Comment by pakrym** [comment]
> not sure that I like having a generic TokenCredential that can't be identified specifically as a static token credential.

Why not?

**Comment by schaabs** [comment]
> What about a static method? TokenCredential.Create(string token)?

I do like this from a design perspective, but honestly I'm not sold on adding the class or the static method. I think either lead users down a bad path because while simply using a static token seems like the path of least resistance but it will only work for a short period of time and the user need to manage it's lifetime and scope properly which is not trivial. This is why we have be reluctant to add this class, and I think that a static method would exacerbate this problem as users could look at token credential (which is abstract so no ctor) find this static method and assume that's what they are intended to use for authentication.

File: sdk/identity/Azure.Identity/samples/DefiningCustomCredentialTypes.md [hunk 1]
```diff
@@ -2,40 +2,14 @@
...
-
-    public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
-    {
-        return new ValueTask<AccessToken>(_token);
-    }
-}
-```
-
-Once the application has defined this credential type instances of it can be used to authenticate Azure SDK clients. The following example shows an how an application already using some other mechanism for acquiring tokens (in this case the hypothetical method `AquireTokenForScope`) could use the `StaticTokenCredential` to authenticate a `BlobClient`.
+The following example shows an how an application already using some other mechanism for acquiring tokens (in this case the hypothetical method `AquireTokenForScope`) could use the `StaticTokenCredential` to authenticate a `BlobClient`.
```
Comment (jongio) [inline sdk/identity/Azure.Identity/samples/DefiningCustomCredentialTypes.md:8]: This needs to be updated right?

File: sdk/identity/Azure.Identity/samples/DefiningCustomCredentialTypes.md [hunk 2]
```diff
@@ -2,40 +2,14 @@
...
 
 ```C# Snippet:StaticTokenCredentialUsage
```
Comment (jongio) [inline sdk/identity/Azure.Identity/samples/DefiningCustomCredentialTypes.md:10]: Probably good to change the snippet name too?

File: sdk/core/Azure.Core/src/TokenCredential.cs [hunk 3]
```diff
@@ -29,5 +30,86 @@ public abstract class TokenCredential
...
+        /// <remarks>
+        /// Typically, the <see cref="TokenCredential"/> created by this method is for use when you have already obtained an <see cref="AccessToken"/>
+        /// from some other source and need a <see cref="TokenCredential"/> that will simply return that token. Because the static token can expire,
+        /// the delegates offer a mechanism to handle <see cref="AccessToken"/> renewal.
+        /// </remarks>
+        /// <param name="getTokenAsync">A delegate that returns a <see cref="ValueTask"/> of type <see cref="AccessToken"/>.
+        /// If <see cref="GetToken"/> is called, the provided delegate will be executed sync over async.</param>
+        /// <returns></returns>
+        public static TokenCredential Create(
+            Func<TokenRequestContext, CancellationToken, ValueTask<AccessToken>> getTokenAsync) => new StaticTokenCredential(getTokenAsync);
```
Comment (schaabs) [inline sdk/core/Azure.Core/src/TokenCredential.cs:74]: Does the func need to return `ValueTask`? If not I wonder if `Task<AccessToken>` would be easier for users.

**Review by schaabs — APPROVED** [review]

File: sdk/core/Azure.Core/src/TokenCredential.cs (hunk 3 above)
Comment (christothes) [reply sdk/core/Azure.Core/src/TokenCredential.cs:74]: I wanted to keep it 1:1 with GetToken's signature, unless you feel strongly otherwise

**Review by pakrym — APPROVED** [review]

File: sdk/core/Azure.Core/src/TokenCredential.cs (hunk 3 above)
Comment (schaabs) [reply sdk/core/Azure.Core/src/TokenCredential.cs:74]: I don't feel strongly I just wanted to pose the question to others.

**Pull request merged**